process/block/shardblock.go

```diff
@@ -679,11 +689,84 @@ func (sp *shardProcessor) CommitBlock(
 		sp.dataPool,
 	)
 
-	sp.blockSizeThrottler.Succeed(uint64(header.Round))
+	sp.blockSizeThrottler.Succeed(header.Round)
 
 	return nil
 }
 
+// getHighestHdrForOwnShardFromMetachain calculates the highest shard header notarized by metachain
+func (sp *shardProcessor) getHighestHdrForOwnShardFromMetachain(round uint64) (*block.Header, []byte, error) {
+	highestNonceOwnShIdHdr := &block.Header{}
+	highestNonceOwnShIdHdrHash, _ := core.CalculateHash(sp.marshalizer, sp.hasher, highestNonceOwnShIdHdr)
+
+	orderedMetaBlocks, err := sp.getOrderedMetaBlocks(round)
+	if err != nil {
+		return highestNonceOwnShIdHdr, highestNonceOwnShIdHdrHash, err
```

Comment:
Consider this scenario: Something errors, we return `&block.Header{}`.
Then, we call `AddHeader` into the fork detector with this empty header as final.

In `AddHeader` we reach the **else** case and will `setFinalCheckpoint` with nonce 0 and round 0.

Reply:
Actually this will not happen. This is why there is on else the condition if finalHeader.GetNonce() > core.GenesisBlockNonce

Reply:
As I wanted to keep the ForkDetector functional for both (metachain and shardchain), and for metachain the final block is calculated inside the metachain shard, than for metachain finalHeader and finalHash will be set to nil on AddHeader calls. But for shardchain it should be send something not nil, this is why it is created the &block.Header{} always. If there was an error the nonce is 0 and this will not set the final as 0 is the nonce of the genesis block and this block could not be received by network as everybody creates it. So, only in the case of a header with nonce greater than 0 the final will be set.

Reply:
Ok, got it.